feat: storage layout + read/write slots + mut keyword

0xrusowsky opened the pull request:
- enforces `mut` keyword for write operations, so that the compiler complains when a contract fn tries to update storage without `&mut self`
- implements wrapper type `Slot<V>` for primitives that fit a single storage slot (equivalent to the old `Word<V>`   >>>   as discussed, we don't like this wrapper type, so i think that down the line we can enhance the `contract` macro so that the devex is like working with primitives.
- implements a new proc macro `storage` that initializes the storage layout >>> it requires all the attributes of the struct to implement the `StorageLayout` trait (some notes regarding this point):
   - maybe the `StorageLayout` trait is unnecessary and it could be merged with the existing `StorageStorable`.
   - i still haven't thought how we will deal with constants (since they should be retrieved from the bytecode), but probably we can change the slot id from `u64` to `Option<u64>` or something like this.
   - we could probably enhance this macro to autogenerate getter fns for all of its attributes
   - i will add tests to validate that the slots are properly initialized and that there is no storage collisions once the design is validated
 
   
PS: i know i'm not the best at coming up with good names, so feel free to propose new ones

leonardoalt commented on erc20/src/lib.rs:
```diff
@@ -3,22 +3,22 @@
-    name: String,
-    symbol: String,
-    decimals: u8,
+    total_supply: Slot<U256>,
```
We chatted about some of this offline, I think it's a bit odd to have `Slot` here but not in Mapping, but good for a first version and can be improved in the future.

leonardoalt replied:
So if I understand correctly, every type used in a struct that derives `storage` has to implement `StorageLayout`, right?

0xrusowsky replied:
> We chatted about some of this offline, I think it's a bit odd to have `Slot` here but not in Mapping, but good for a first version and can be improved in the future.

yes, my bad. I totally forgot... I'll amend the PR to incorporate that design pattern.

> So if I understand correctly, every type used in a struct that derives `storage` has to implement `StorageLayout`, right?

yes, that's correct they need to impl `StorageLayout` so that we can ensure that we can allocate a slot in the layout for them.

leonardoalt replied:
I think this is fine as is and we can iterate later on

leonardoalt commented on erc20/src/lib.rs:
```diff
@@ -81,25 +98,17 @@ impl ERC20 {
+
     pub fn total_supply(&self) -> U256 {
-        self.total_supply
+        self.total_supply.read()
```
I think you wrote this somewhere, but `pub` fields could also get automatic getters, can be done in the future. (let's open an issue)

leonardoalt commented on eth-riscv-runtime/src/lib.rs:
```diff
@@ -22,8 +22,8 @@ pub use call::call_contract;
-    fn call(&self);
-    fn call_with_data(&self, calldata: &[u8]);
+    fn call(&mut self);
+    fn call_with_data(&mut self, calldata: &[u8]);
```
It could also make sense later on to have immutable versions of these, not sure how that would work but it'd be cool.

0xrusowsky replied:
agreed.
do you foresee any scenario other than static calls where we would use the not mutable call?

leonardoalt replied:
not right now

leonardoalt commented on eth-riscv-runtime/src/types/mapping.rs:
```diff
@@ -1,52 +1,30 @@
-        padded[..bytes.len()].copy_from_slice(&bytes);
-        sstore(key, U256::from_be_bytes(padded));
+impl<K, V> StorageLayout for Mapping<K, V> {
+    fn allocate(slot: u64) -> Self {
```
Nice!

leonardoalt commented on eth-riscv-runtime/src/types/slot.rs:
```diff
@@ -0,0 +1,77 @@
+/// Wrapper around `alloy::primitives` that can be written in a single slot (single EVM word).
+#[derive(Default)]
+pub struct Slot<V> {
+    id: u64,
```
Technically, shouldn't the id  be u256? I agree it should be fine as u64 if we incrementally allocate slots, but to be fully compatible/flexible and allow custom storage layouts from the user I think this would have to be u256?

0xrusowsky replied:
yes, it should be u256, i just kept the u64 as that's what was being used in `Mapping`. I'll do a follow-up commit to make everything compliant.

leonardoalt commented on contract-derive/src/lib.rs:
```diff
@@ -326,3 +326,49 @@ pub fn interface(_attr: TokenStream, item: TokenStream) -> TokenStream {
+    });
+
+    // Generate initialization code for each field
+    let init_fields = fields.iter().enumerate().map(|(slot, f)| {
```
I think this is fine for now, but we should spend some time crafting how storage layout will work. For example, this won't allow for storing a `(u256, u256)` sequentially, we'd need to implement `StorageLayout` for it with a phantom slot, and then in the actual read/write use some encoded key (similar to Mapping). Ideally we'd know the storage size of the type at compile time (if value type) and use that size for allocation instead of `enumerate`. Reference types (`Mapping`, arrays, etc) could work like Mapping. This is what Solidity does also.

0xrusowsky replied:
This is a really good point and i totally missed that. Do you want me to try to address it in this PR? or would u open an issue and do it in the following one?

leonardoalt replied:
I think later is fine

0xrusowsky commented:
@leonardoalt changes since your review:
- as suggested, made it EVM compliant by changing from a u64-based storage layout to a u256 one.
- i've also changed little endian usage to big endian (in the execution side only, i left the read/writes from riscv memory as they were) to follow the EVM compliant directive.
- added tests to validate that the generated storage layout is correct.

finally, i tracked the todo's here: https://github.com/r55-eth/r55/issues/32

leonardoalt reviewed (commented):
Just one small comment

0xrusowsky commented on README.md:
```diff
@@ -37,7 +37,7 @@ out-of-the-box.
 
 use contract_derive::contract;
-use eth_riscv_runtime::types::Mapping;
+use eth_riscv_runtime::storage::Mapping;
```
sure, i just thought `storage` would be more informative.

we can move back to a more generic `types`

0xrusowsky commented on eth-riscv-runtime/src/storage/mod.rs:
```diff
@@ -15,26 +15,27 @@ mod slot;
+// TODO: enhance `storage` macro to handle complex types (like tuples or custom structs)
 pub trait StorageLayout {
-    fn allocate(slot: u64) -> Self;
+    fn allocate(limb0: u64, limb1: u64, limb2: u64, limb3: u64) -> Self;
```
same reason as above

0xrusowsky commented on eth-riscv-runtime/src/storage/mapping.rs:
```diff
@@ -3,38 +3,38 @@ use super::*;
 
 impl<K, V> StorageLayout for Mapping<K, V> {
-    fn allocate(slot: u64) -> Self {
+    fn allocate(first: u64, second: u64, third: u64, fourth: u64) -> Self {
```
otherwise i can maybe add comments for it to make it more comprehensive

leonardoalt replied:
ah right
